File: core/src/main/java/com/andiag/core/BaseTabsFragment.java
package com.andiag.core;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.TabLayout;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentPagerAdapter;
import android.support.v4.view.ViewPager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.util.ArrayList;
import java.util.List;

public abstract class BaseTabsFragment<P extends BasePresenter> extends BaseFragment<P> {

    private static final String SELECTED_TAB = "SELECTED_TAB";

    ViewPager mViewPager;
    TabLayout mTabLayout;

    private int mSelectedTab = 0;

    @Override
    protected int getLayout() {
        return R.layout.fragment_tabs;
    }

    private void setupTabs() {
        TabsAdapter adapter = new TabsAdapter(getChildFragmentManager());
        List<BaseTab> tabs = getTabs();
        for (BaseTab tab : tabs) {
            adapter.addTab(tab);
        }
        mViewPager.setAdapter(adapter);
        mTabLayout.setupWithViewPager(mViewPager);
        setupTabLayout(tabs);
    }

    private void setupTabLayout(List<BaseTab> tabs) {
        for (int i = 0; i < mTabLayout.getTabCount(); i++) {
            mTabLayout.getTabAt(i).setIcon(tabs.get(i).getIcon());
        }
    }

    public abstract List<BaseTab> getTabs();

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putInt(SELECTED_TAB, mSelectedTab);
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (savedInstanceState != null && savedInstanceState.containsKey(SELECTED_TAB)) {
            mSelectedTab = savedInstanceState.getInt(SELECTED_TAB);
        }
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = super.onCreateView(inflater, container, savedInstanceState);
        mViewPager = view.findViewById(R.id.viewPager);
        mTabLayout = view.findViewById(R.id.tabLayout);

        setupTabs();

        return view;
    }

    @Override
    public void onResume() {
        super.onResume();
        mViewPager.setCurrentItem(mSelectedTab);
    }

    @Override
    public void onStart() {
        super.onStart();

        mViewPager.addOnPageChangeListener(new ViewPager.OnPageChangeListener() {
            @Override
            public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) { }
            @Override
            public void onPageSelected(int position) { mSelectedTab = position; }
            @Override
            public void onPageScrollStateChanged(int state) { }
        });
    }

    @Override
    public void onStop() {
        super.onStop();
        mViewPager.clearOnPageChangeListeners();
    }

    static class TabsAdapter extends FragmentPagerAdapter {
        private final List<Fragment> mFragments = new ArrayList<>();
        private final List<String> mFragmentTitles = new ArrayList<>();

        TabsAdapter(FragmentManager fm) {
            super(fm);
        }

        void addTab(BaseTab tab) {
            mFragments.add(tab.getFragment());
            mFragmentTitles.add(tab.getTitle());
        }

        @Override
        public Fragment getItem(int position) {
            return mFragments.get(position);
        }

        @Override
        public int getCount() {
            return mFragments.size();
        }

        @Override
        public CharSequence getPageTitle(int position) {
            CharSequence pageTitle = mFragmentTitles.get(position);
            return pageTitle.equals("") ? pageTitle : null;
        }
    }
}

File: core/src/main/java/com/andiag/core/BaseTabsFragment.kt
package com.andiag.core

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentPagerAdapter
import androidx.viewpager.widget.ViewPager
import com.google.android.material.tabs.TabLayout
import java.util.*

abstract class BaseTabsFragment<P : BasePresenter> : BaseFragment<P>() {

    lateinit var mViewPager: ViewPager
    lateinit var mTabLayout: TabLayout

    private var mSelectedTab = 0

    abstract val tabs: List<BaseTab>

    override val layout: Int
        get() = R.layout.fragment_tabs

    private fun setupTabs() {
        val adapter = TabsAdapter(childFragmentManager)
        val tabs = tabs
        for (tab in tabs) {
            adapter.addTab(tab)
        }
        mViewPager.adapter = adapter
        mTabLayout.setupWithViewPager(mViewPager)
        setupTabLayout(tabs)
    }

    private fun setupTabLayout(tabs: List<BaseTab>) {
        for (i in 0 until mTabLayout.tabCount) {
            mTabLayout.getTabAt(i)!!.setIcon(tabs[i].icon)
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putInt(SELECTED_TAB, mSelectedTab)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (savedInstanceState != null && savedInstanceState.containsKey(SELECTED_TAB)) {
            mSelectedTab = savedInstanceState.getInt(SELECTED_TAB)
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = super.onCreateView(inflater, container, savedInstanceState)
        mViewPager = view!!.findViewById(R.id.viewPager)
        mTabLayout = view.findViewById(R.id.tabLayout)

        setupTabs()

        return view
    }

    override fun onResume() {
        super.onResume()
        mViewPager.currentItem = mSelectedTab
    }

    override fun onStart() {
        super.onStart()

        mViewPager.addOnPageChangeListener(object : ViewPager.OnPageChangeListener {
            override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {}
            override fun onPageSelected(position: Int) {
                mSelectedTab = position
            }

            override fun onPageScrollStateChanged(state: Int) {}
        })
    }

    override fun onStop() {
        super.onStop()
        mViewPager.clearOnPageChangeListeners()
    }

    internal class TabsAdapter(fm: FragmentManager) : FragmentPagerAdapter(fm) {
        private val mFragments = ArrayList<Fragment>()
        private val mFragmentTitles = ArrayList<String>()

        fun addTab(tab: BaseTab) {
            mFragments.add(tab.fragment)
            mFragmentTitles.add(tab.title)
        }

        override fun getItem(position: Int): Fragment {
            return mFragments[position]
        }

        override fun getCount(): Int {
            return mFragments.size
        }

        override fun getPageTitle(position: Int): CharSequence? {
            val pageTitle = mFragmentTitles[position]
            return if (pageTitle == "") pageTitle else null
        }
    }

    companion object {
        private val SELECTED_TAB = "SELECTED_TAB"
    }
}
